common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java

```diff
@@ -436,11 +440,27 @@ public static ClientException clientExceptionFromException(@NonNull final Throwa
 
 
         if (e instanceof OutOfMemoryError) {
-            return new ClientException(
-                    ClientException.OUT_OF_MEMORY,
-                    e.getMessage(),
-                    e
-            );
+            // When receiving an out of memory error, instead of gracefully returning a failure result, we should shut down
+            // current broker process, to allow a new broker process to be launched at the next request from client app.
+            // This will result in a new broker process with fresh memory allocation, rather than repeating out of memory
+            // errors. In the case that this is an msal-only scenario, and this code is running inside client app, then
+            // client app will be shut down and will need to be launched again.
+
+            // Log status code and record exception in span
+            final Span currentSpan = SpanExtension.current();
+            currentSpan.setStatus(StatusCode.ERROR);
+            currentSpan.recordException(e);
+
+            // Attach the stack trace, to debug in telemetry later
+            currentSpan.setAttribute(AttributeName.out_of_memory_exception_stacktrace.name(),
+                    StringUtil.getStacktraceAsStringFromElementArray(e.getStackTrace()));
+
+            Logger.error(TAG, "Received an out of memory error, shutting broker process so a new one can be launched.", e);
+
+            // Shut down current process
+            // Calling client app will receive "killed unexpectedly" error from MSAL if broker is shut down with this statement
+            // Calling application should not crash, and should be able to make another call, which will result in a new broker process.
+            System.exit(0);
```

Comment:
What error/response does calling app receive in when this happens?

Reply:
There's also android.os.Process.killProcess(android.os.Process.myPid());  this might be better to use in case of OOM. I read a bit, you may also read once and we can finalize

Reply:
Off of a quick glance, it seems using system.exit is preferable because it shuts down the JVM itself, while killprocess only kills the android process. I think from a memory management perspective, wouldn't shutting down the JVM be preferable?

Reply:
To answer your question, we get an MSALException, Activity is killed unexpectedly

![image](https://github.com/user-attachments/assets/b5a357d5-7005-4485-a4e4-1f5a3203bf81)

Reply:
I read otherwise on the internet that, prefer killprocess to completely release all resources jvm and no-jvm, and all threads

Reply:
Do you same error when the OOM happened in acquire token silent case?

Reply:
Same thing happens for Silent. Issue with using killprocess is that our exceptionAdapter is in common4j. I could look at creating a wrapper in android for exceptionAdapter specifically for this change but it seems like overkill, not sure the advantage of using killprocess justifies that kind of change.

Reply:
I've reworked to instead handle this in command dispatcher using IPlatformUtil